File: Stiver_Problems/Graphs/P011.cpp
// P011.cpp - Remove stones (max stones removed)

#include <bits/stdc++.h>
using namespace std;

// Brute: build graph of stones sharing row/col, count connected components
int brute(const vector<pair<int,int>>& stones) {
    int n = stones.size();
    vector<vector<int>> adj(n);
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            if (stones[i].first == stones[j].first ||
                stones[i].second == stones[j].second) {
                adj[i].push_back(j);
                adj[j].push_back(i);
            }
        }
    }
    vector<bool> visited(n, false);
    function<void(int)> dfs = [&](int node) {
        visited[node] = true;
        for (int nei : adj[node]) {
            if (!visited[nei]) dfs(nei);
        }
    };

    int comp = 0;
    for (int i = 0; i < n; i++) {
        if (!visited[i]) {
            dfs(i);
            comp++;
        }
    }
    return n - comp;
}

// DSU for optimal solution
struct DSU {
    vector<int> parent, rankv;
    DSU(int n) : parent(n), rankv(n, 1) {
        iota(parent.begin(), parent.end(), 0);
    }
    int find(int u) {
        if (parent[u] == u) return u;
        return parent[u] = find(parent[u]);
    }
    void unite(int u, int v) {
        int pu = find(u), pv = find(v);
        if (pu == pv) return;
        if (rankv[pu] > rankv[pv]) parent[pv] = pu;
        else if (rankv[pv] > rankv[pu]) parent[pu] = pv;
        else {
            parent[pu] = pv;
            rankv[pv]++;
        }
    }
};

// Optimal: DSU on rows + columns
int optimal(const vector<pair<int,int>>& stones) {
    int n = stones.size();
    int maxRow = 0, maxCol = 0;
    for (auto &p : stones) {
        maxRow = max(maxRow, p.first);
        maxCol = max(maxCol, p.second);
    }
    int offset = maxRow + 1;
    DSU dsu(maxRow + maxCol + 2);

    for (auto &p : stones) {
        int r = p.first;
        int c = p.second + offset;
        dsu.unite(r, c);
    }

    unordered_set<int> roots;
    for (auto &p : stones) {
        roots.insert(dsu.find(p.first));
    }
    int C = (int)roots.size();
    return n - C;
}

int main() {
    vector<pair<int,int>> stones = {
        {0,0},{0,1},{1,0},{1,2},{2,2},{2,3}
    };
    cout << "Brute Force: " << brute(stones) << "\n";
    cout << "Optimal DSU: " << optimal(stones) << "\n";
    return 0;
}

File: Graphs/P011.py
def brute(stones):
    n=len(stones)
    adj=[[] for _ in range(n)]
    for i in range(n):
        for j in range(i+1,n):
            if stones[i][0]==stones[j][0] or stones[i][1]==stones[j][1]:
                adj[i].append(j)
                adj[j].append(i)
    visited=[False]*n
    def dfs(node):
        visited[node]=True
        for nei in adj[node]:
            if not visited[nei]:
                dfs(nei)
    C=[0]
    for i in range(n):
        if not visited[i]:
            dfs(i)
            C[0]+=1
    return n-C[0]


class DSU:
    def __init__(self,n):
        self.parent=[i for i in range(n)]
        self.rank=[1]*n
    def find(self,u):
        if self.parent[u]==u:
            return u
        self.parent[u]=self.find(self.parent[u])
        return self.parent[u]
    def union(self,u,v):
        pu=self.find(u)
        pv=self.find(v)
        if self.rank[pu]>self.rank[pv]:
            self.parent[pv]=pu
        elif self.rank[pu]<self.rank[pv]:
            self.parent[pu]=pv
        else:
            self.parent[pv]=pu
            self.rank[pu]+=1

def optimal(stones):
    n=len(stones)
    row=col=0
    for r,c in stones:
        row=max(row,r)
        col=max(col,c)
    offset=row+1
    dsu=DSU(row+col+2)
    for r,c in stones:
        dsu.union(r,c+offset)
    C=len(set(dsu.find(r) for r,c in stones))
    return n-C

if __name__ == "__main__":
    stones = [[0,0],[0,1],[1,0],[1,2],[2,2],[2,3]]
    print("Brute Force:", brute(stones))
    print("Optimal DSU:", optimal(stones))